roscala/src/main/scala/coop/rchain/rosette/VirtualMachine.scala

```diff
@@ -351,19 +371,43 @@ object VirtualMachine {
     state.update(_ >> 'ctxt)(Ctxt(Some(Tuple(op.n, None)), _))
 
   def execute(op: OpExtend, state: VMState): VMState = {
-    val formals = state.code.lit(op.v).asInstanceOf[Template]
-    val actuals = formals.matchPattern(state.ctxt.argvec, state.ctxt.nargs)
+    val formals = state.code.lit(op.v).as[Template]
+
+    formals match {
+      case Some(template) =>
+        // tries to match formals to actuals
+        val actuals: Option[Tuple] =
+          template.`match`(state.ctxt.argvec, state.ctxt.nargs)
+
+        actuals match {
+          // extend the current environment with the actuals
+          // reset state.ctxt.nargs to 0
+          case Some(tuple) =>
+            state.ctxt.env.as[StdExtension] match {
+              case Some(env) =>
+                env.extendWith(template.keyMeta, tuple) match {
+                  case Some(newEnv) =>
+                    state
+                      .set(_ >> 'ctxt >> 'env)(newEnv)
+                      .set(_ >> 'ctxt >> 'nargs)(0)
+
+                  case None =>
+                    die("OpExtend: could not extend state.ctxt.env")(state)
+                }
+
+              case None =>
+                die("OpExtend: state.ctxt.env needs to be a StdExtension")(
+                  state)
+            }
 
-    actuals match {
-      case Some(tuple) =>
-        state
-          .set(_ >> 'ctxt >> 'nargs)(0)
-          .set(_ >> 'ctxt >> 'env)(
-            state.ctxt.env.extendWith(formals.keymeta, tuple))
+          case None =>
+            // TODO: Revisit
+            //handleFormalsMismatch(formals)
+            state.set(_ >> 'doNextThreadFlag)(true)
+        }
```

Comment:
Personally, I would prefer to avoid explicit matching and use something like this instead:
```scala
import cats.implicits._

val a: Option[Int] = None
val b: Option[Int] = Some(2)
val c: Option[Int] = Some(3)

implicit class OptionOps[R](opt: Option[R]) {
  def or[L](alt: L): Either[L,R] = opt.toRight(alt) 
}

for {
  aa <- a or "hello"
  bb <- b or "world"
  cc <- c or "here"
} yield aa + bb + cc
```
I'm using implicit class here just for fancy syntax.

Reply:
Pretty neat!